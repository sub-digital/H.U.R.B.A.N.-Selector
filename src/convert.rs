use std::convert::TryInto;
use std::fmt::Debug;

/// Convert `n` to `u32` using `TryFrom` or panic.
///
/// # Panics
/// Panics if the conversion returns an error.
pub fn cast_u32<T>(n: T) -> u32
where
    T: TryInto<u32>,
    <T as TryInto<u32>>::Error: Debug,
{
    n.try_into().expect("Expected N to fit in u32")
}

/// Convert `n` to `i32` using `TryFrom` or panic.
///
/// # Panics
/// Panics if the conversion returns an error.
pub fn cast_i32<T>(n: T) -> i32
where
    T: TryInto<i32>,
    <T as TryInto<i32>>::Error: Debug,
{
    n.try_into().expect("Expected N to fit in i32")
}

/// Convert `n` to `usize` using `TryFrom` or panic.
///
/// # Panics
/// Panics if the conversion returns an error.
pub fn cast_usize<T>(n: T) -> usize
where
    T: TryInto<usize>,
    <T as TryInto<usize>>::Error: Debug,
{
    n.try_into().expect("Expected N to fit in usize")
}

/// Convert u32 to i32 clamping to max value of i32 if necessary.
pub fn clamp_cast_u32_to_i32(n: u32) -> i32 {
    if n > i32::max_value() as u32 {
        i32::max_value()
    } else {
        n as i32
    }
}

/// Convert i32 to u32 clamping to 0 if necessary.
pub fn clamp_cast_i32_to_u32(n: i32) -> u32 {
    if n < 0 {
        0
    } else {
        n as u32
    }
}

/// Convert u8 based color array to f32 based color array with
/// floating point domain `[0..1]`.
pub fn cast_u8_color_to_f32(color: [u8; 4]) -> [f32; 4] {
    [
        color[0] as f32 / 255.0,
        color[1] as f32 / 255.0,
        color[2] as f32 / 255.0,
        color[3] as f32 / 255.0,
    ]
}

/// Convert u8 based color array to f64 based color array with
/// floating point domain `[0..1]`.
pub fn cast_u8_color_to_f64(color: [u8; 4]) -> [f64; 4] {
    [
        color[0] as f64 / 255.0,
        color[1] as f64 / 255.0,
        color[2] as f64 / 255.0,
        color[3] as f64 / 255.0,
    ]
}
