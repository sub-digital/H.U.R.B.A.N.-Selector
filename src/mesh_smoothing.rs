use std::collections::HashMap;

use nalgebra::geometry::Point3;
use smallvec::SmallVec;

use crate::convert::cast_usize;
use crate::geometry::Geometry;

/// Relaxes angles between mesh edges, resulting in a smoother geometry
///
/// or
///
/// Relaxes angles between mesh edges, while optionally keeping some vertices
/// anchored, resulting in an evenly distributed geometry optionally stretched
/// between the anchor points
///
/// The number of vertices, faces and the overall topology remains unchanged.
/// The more iterations, the smoother result. Too many iterations may cause slow
/// calculation time. In case the stop_when_stable flag is set on, the smoothing
/// stops when the geometry stops transforming between iterations or when it
/// reaches the maximum number of iterations.
///
/// The algorithm is based on replacing each vertex position with an average
/// position of its immediate neighbors.
///
/// - `geometry` - mesh geometry to relax
/// - `iterations` - (maximum) number of times the smoothing algorithm should
///   relax the geometry
/// - `fixed_vertex_indices` - indices of vertices to keep fixed during the
///   relaxation
/// - `stop_when_stable` - the smoothing stops when there is no change between
///   iterations
///
/// returns (smooth_geometry: Geometry, executed_iterations: u32, stable: bool)
pub fn laplacian_smoothing(
    geometry: &Geometry,
    vertex_to_vertex_topology: &HashMap<u32, SmallVec<[u32; 8]>>,
    iterations: u32,
    fixed_vertex_indices: &[u32],
    stop_when_stable: bool,
) -> (Geometry, u32, bool) {
    if iterations == 0 {
        return (geometry.clone(), 0, false);
    }

    let mut vertices: Vec<Point3<f32>> = Vec::from(geometry.vertices());
    let mut geometry_vertices: Vec<Point3<f32>>;

    let mut iteration: u32 = 0;

    // Only relevant when fixed vertices are specified
    let mut stable = !fixed_vertex_indices.is_empty();
    while iteration < iterations {
        stable = !fixed_vertex_indices.is_empty();
        geometry_vertices = vertices.clone();

        for (current_vertex_index, neighbors_indices) in vertex_to_vertex_topology.iter() {
            if fixed_vertex_indices
                .iter()
                .all(|i| i != current_vertex_index)
                && !neighbors_indices.is_empty()
            {
                let mut average_position: Point3<f32> = Point3::origin();
                for neighbor_index in neighbors_indices {
                    average_position += geometry_vertices[cast_usize(*neighbor_index)].coords;
                }
                average_position /= neighbors_indices.len() as f32;
                stable &= approx::relative_eq!(
                    &average_position.coords,
                    &vertices[cast_usize(*current_vertex_index)].coords,
                );
                vertices[cast_usize(*current_vertex_index)] = average_position;
            }
        }
        iteration += 1;

        if stop_when_stable && stable {
            break;
        }
    }

    // FIXME: Calculate smooth normals for the result once we support them
    (
        Geometry::from_faces_with_vertices_and_normals(
            geometry.faces().to_vec(),
            vertices,
            geometry.normals().to_vec(),
        ),
        iteration,
        stable,
    )
}

#[cfg(test)]
mod tests {
    use std::iter::FromIterator;

    use nalgebra;

    use crate::edge_analysis;
    use crate::geometry::{Geometry, NormalStrategy, OrientedEdge, Vertices};
    use crate::mesh_analysis;
    use crate::mesh_topology_analysis;

    use super::*;

    // FIXME: Snapshot testing
    fn torus() -> (Vec<(u32, u32, u32)>, Vertices) {
        let vertices = vec![
            Point3::new(0.566987, -1.129e-11, 0.25),
            Point3::new(-0.716506, 1.241025, 0.25),
            Point3::new(-0.283494, 0.491025, 0.25),
            Point3::new(-0.716506, -1.241025, 0.25),
            Point3::new(-0.283494, -0.491025, 0.25),
            Point3::new(1.0, -1.129e-11, -0.5),
            Point3::new(1.433013, -1.129e-11, 0.25),
            Point3::new(-0.5, 0.866025, -0.5),
            Point3::new(-0.5, -0.866025, -0.5),
        ];

        let faces = vec![
            (4, 3, 6),
            (0, 6, 2),
            (2, 1, 3),
            (8, 4, 0),
            (3, 8, 6),
            (5, 0, 7),
            (6, 5, 7),
            (7, 2, 4),
            (1, 7, 8),
            (4, 6, 0),
            (6, 1, 2),
            (2, 3, 4),
            (8, 0, 5),
            (8, 5, 6),
            (0, 2, 7),
            (6, 7, 1),
            (7, 4, 8),
            (1, 8, 3),
        ];

        (faces, vertices)
    }

    fn triple_torus() -> (Vec<(u32, u32, u32)>, Vertices) {
        let vertices = vec![
            Point3::new(15.566987, -1.129e-11, 0.25),
            Point3::new(14.283494, 1.241025, 0.25),
            Point3::new(14.716506, 0.491025, 0.25),
            Point3::new(14.283494, -1.241025, 0.25),
            Point3::new(14.716506, -0.491025, 0.25),
            Point3::new(16.0, 0.75, 0.25),
            Point3::new(15.149519, 1.241025, 0.25),
            Point3::new(16.0, 1.732051, 0.25),
            Point3::new(16.108253, 0.1875, -0.5),
            Point3::new(16.433012, -1.129e-11, 0.25),
            Point3::new(14.716506, 1.991025, 0.25),
            Point3::new(15.566987, 2.482051, 0.25),
            Point3::new(14.283494, 3.723076, 0.25),
            Point3::new(14.716506, 2.973076, 0.25),
            Point3::new(14.554127, 1.334775, -0.5),
            Point3::new(14.5, -0.866025, -0.5),
            Point3::new(14.5, 3.348076, -0.5),
            Point3::new(16.108253, 2.294551, -0.5),
            Point3::new(16.433012, 2.482051, 0.25),
        ];

        let faces = vec![
            (4, 3, 0),
            (0, 9, 1),
            (2, 1, 3),
            (7, 5, 9),
            (5, 6, 9),
            (6, 7, 18),
            (15, 4, 0),
            (3, 15, 9),
            (10, 1, 11),
            (11, 18, 12),
            (13, 12, 1),
            (14, 2, 15),
            (1, 14, 15),
            (8, 0, 2),
            (8, 14, 6),
            (16, 13, 10),
            (12, 16, 1),
            (17, 8, 7),
            (18, 9, 8),
            (14, 17, 6),
            (17, 11, 16),
            (18, 17, 16),
            (14, 10, 17),
            (3, 9, 0),
            (0, 1, 2),
            (2, 3, 4),
            (7, 9, 18),
            (6, 1, 9),
            (6, 18, 1),
            (15, 0, 8),
            (15, 8, 9),
            (1, 18, 11),
            (11, 12, 13),
            (13, 1, 10),
            (2, 4, 15),
            (1, 15, 3),
            (8, 2, 14),
            (8, 6, 5),
            (16, 10, 14),
            (16, 14, 1),
            (8, 5, 7),
            (18, 8, 17),
            (17, 7, 6),
            (11, 13, 16),
            (18, 16, 12),
            (10, 11, 17),
        ];

        (faces, vertices)
    }

    fn shape_for_smoothing_with_anchors() -> (Vec<(u32, u32, u32)>, Vertices) {
        let vertices = vec![
            Point3::new(30.21796, -6.119943, 0.0),
            Point3::new(32.031532, 1.328689, 0.0),
            Point3::new(33.875141, -3.522298, 3.718605),
            Point3::new(34.571838, -2.071111, 2.77835),
            Point3::new(34.778172, -5.285372, 3.718605),
            Point3::new(36.243252, -3.80194, 3.718605),
            Point3::new(36.741604, -10.146505, 0.0),
            Point3::new(39.676025, 1.905633, 0.0),
            Point3::new(42.587009, -5.186427, 0.0),
        ];

        let faces = vec![
            (4, 8, 5),
            (4, 6, 8),
            (5, 8, 7),
            (3, 5, 7),
            (0, 2, 1),
            (1, 2, 3),
            (0, 4, 2),
            (1, 3, 7),
            (0, 6, 4),
            (2, 4, 5),
            (2, 5, 3),
        ];

        (faces, vertices)
    }

    fn shape_for_smoothing_with_anchors_50_iterations() -> (Vec<(u32, u32, u32)>, Vertices) {
        let vertices = vec![
            Point3::new(30.21796, -6.119943, 0.0),
            Point3::new(32.031532, 1.328689, 0.0),
            Point3::new(34.491065, -2.551039, 0.0),
            Point3::new(36.00632, -0.404003, 0.0),
            Point3::new(36.372859, -5.260642, 0.0),
            Point3::new(37.826656, -2.299296, 0.0),
            Point3::new(36.741604, -10.146505, 0.0),
            Point3::new(39.676025, 1.905633, 0.0),
            Point3::new(42.587009, -5.186427, 0.0),
        ];

        let faces = vec![
            (4, 8, 5),
            (4, 6, 8),
            (5, 8, 7),
            (3, 5, 7),
            (0, 2, 1),
            (1, 2, 3),
            (0, 4, 2),
            (1, 3, 7),
            (0, 6, 4),
            (2, 4, 5),
            (2, 5, 3),
        ];

        (faces, vertices)
    }

    #[test]
    fn test_laplacian_smoothing_preserves_face_vertex_normal_count() {
        let (faces, vertices) = torus();
        let geometry = Geometry::from_triangle_faces_with_vertices_and_computed_normals(
            faces.clone(),
            vertices.clone(),
            NormalStrategy::Sharp,
        );

        let vertex_to_vertex_topology =
            mesh_topology_analysis::vertex_to_vertex_topology(&geometry);
        let (relaxed_geometry_0, _, _) =
            laplacian_smoothing(&geometry, &vertex_to_vertex_topology, 0, &[], false);
        let (relaxed_geometry_1, _, _) =
            laplacian_smoothing(&geometry, &vertex_to_vertex_topology, 1, &[], false);
        let (relaxed_geometry_10, _, _) =
            laplacian_smoothing(&geometry, &vertex_to_vertex_topology, 10, &[], false);

        assert_eq!(relaxed_geometry_0.faces().len(), geometry.faces().len(),);
        assert_eq!(relaxed_geometry_1.faces().len(), geometry.faces().len(),);
        assert_eq!(relaxed_geometry_10.faces().len(), geometry.faces().len(),);
        assert_eq!(
            relaxed_geometry_0.vertices().len(),
            geometry.vertices().len(),
        );
        assert_eq!(
            relaxed_geometry_1.vertices().len(),
            geometry.vertices().len(),
        );
        assert_eq!(
            relaxed_geometry_10.vertices().len(),
            geometry.vertices().len(),
        );
        assert_eq!(relaxed_geometry_0.normals().len(), geometry.normals().len());
        assert_eq!(relaxed_geometry_1.normals().len(), geometry.normals().len());
        assert_eq!(
            relaxed_geometry_10.normals().len(),
            geometry.normals().len(),
        );
    }

    #[test]
    fn test_laplacian_smoothing_preserves_original_geometry_with_0_iterations() {
        let (faces, vertices) = triple_torus();
        let geometry = Geometry::from_triangle_faces_with_vertices_and_computed_normals(
            faces,
            vertices,
            NormalStrategy::Sharp,
        );
        let v2v = mesh_topology_analysis::vertex_to_vertex_topology(&geometry);

        let (relaxed_geometry, _, _) = laplacian_smoothing(&geometry, &v2v, 0, &[], false);
        assert_eq!(geometry, relaxed_geometry);
    }

    #[test]
    fn test_laplacian_smoothing_snapshot_triple_torus_1_iteration() {
        let (faces, vertices) = triple_torus();
        let geometry = Geometry::from_triangle_faces_with_vertices_and_computed_normals(
            faces,
            vertices,
            NormalStrategy::Sharp,
        );
        let v2v = mesh_topology_analysis::vertex_to_vertex_topology(&geometry);

        let (relaxed_geometry, _, _) = laplacian_smoothing(&geometry, &v2v, 1, &[], false);
        insta::assert_json_snapshot!(
            "triple_torus_after_1_iteration_of_laplacian_smoothing",
            &relaxed_geometry
        );
    }

    #[test]
    fn test_laplacian_smoothing_snapshot_triple_torus_2_iterations() {
        let (faces, vertices) = triple_torus();
        let geometry = Geometry::from_triangle_faces_with_vertices_and_computed_normals(
            faces,
            vertices,
            NormalStrategy::Sharp,
        );
        let v2v = mesh_topology_analysis::vertex_to_vertex_topology(&geometry);

        let (relaxed_geometry, _, _) = laplacian_smoothing(&geometry, &v2v, 2, &[], false);
        insta::assert_json_snapshot!(
            "triple_torus_after_2_iteration2_of_laplacian_smoothing",
            &relaxed_geometry
        );
    }

    #[test]
    fn test_laplacian_smoothing_snapshot_triple_torus_3_iterations() {
        let (faces, vertices) = triple_torus();
        let geometry = Geometry::from_triangle_faces_with_vertices_and_computed_normals(
            faces,
            vertices,
            NormalStrategy::Sharp,
        );
        let v2v = mesh_topology_analysis::vertex_to_vertex_topology(&geometry);

        let (relaxed_geometry, _, _) = laplacian_smoothing(&geometry, &v2v, 3, &[], false);
        insta::assert_json_snapshot!(
            "triple_torus_after_3_iterations_of_laplacian_smoothing",
            &relaxed_geometry
        );
    }

    #[test]
    fn test_laplacian_smoothing_with_anchors() {
        let (faces, vertices) = shape_for_smoothing_with_anchors();
        let geometry = Geometry::from_triangle_faces_with_vertices_and_computed_normals(
            faces.clone(),
            vertices.clone(),
            NormalStrategy::Sharp,
        );

        let fixed_vertex_indices: Vec<u32> = vec![0, 1, 7, 8, 6];

        let (faces_correct, vertices_correct) = shape_for_smoothing_with_anchors_50_iterations();
        let test_geometry_correct =
            Geometry::from_triangle_faces_with_vertices_and_computed_normals(
                faces_correct.clone(),
                vertices_correct.clone(),
                NormalStrategy::Sharp,
            );

        let v2v = mesh_topology_analysis::vertex_to_vertex_topology(&geometry);
        let (relaxed_geometry, _, _) =
            laplacian_smoothing(&geometry, &v2v, 50, &fixed_vertex_indices, false);

        let relaxed_geometry_faces = relaxed_geometry.faces();
        let test_geometry_faces = test_geometry_correct.faces();

        assert_eq!(relaxed_geometry_faces, test_geometry_faces);

        const TOLERANCE_SQUARED: f32 = 0.01 * 0.01;

        let relaxed_geometry_vertices = relaxed_geometry.vertices();
        let test_geometry_vertices = test_geometry_correct.vertices();

        for i in 0..test_geometry_vertices.len() {
            assert!(
                nalgebra::distance_squared(
                    &test_geometry_vertices[i],
                    &relaxed_geometry_vertices[i]
                ) < TOLERANCE_SQUARED
            );
        }
    }

    #[test]
    fn test_laplacian_smoothing_with_anchors_find_border_vertices() {
        let (faces, vertices) = shape_for_smoothing_with_anchors();
        let geometry = Geometry::from_triangle_faces_with_vertices_and_computed_normals(
            faces.clone(),
            vertices.clone(),
            NormalStrategy::Sharp,
        );

        let oriented_edges: Vec<OrientedEdge> = geometry.oriented_edges_iter().collect();
        let edge_sharing_map = edge_analysis::edge_sharing(&oriented_edges);
        let fixed_vertex_indices =
            Vec::from_iter(mesh_analysis::border_vertex_indices(&edge_sharing_map).into_iter());

        let (faces_correct, vertices_correct) = shape_for_smoothing_with_anchors_50_iterations();
        let test_geometry_correct =
            Geometry::from_triangle_faces_with_vertices_and_computed_normals(
                faces_correct.clone(),
                vertices_correct.clone(),
                NormalStrategy::Sharp,
            );
        let v2v = mesh_topology_analysis::vertex_to_vertex_topology(&geometry);
        let (relaxed_geometry, _, _) =
            laplacian_smoothing(&geometry, &v2v, 50, &fixed_vertex_indices, false);

        let relaxed_geometry_faces = relaxed_geometry.faces();
        let test_geometry_faces = test_geometry_correct.faces();

        assert_eq!(relaxed_geometry_faces, test_geometry_faces);

        let relaxed_geometry_vertices = relaxed_geometry.vertices();
        let test_geometry_vertices = test_geometry_correct.vertices();

        for i in 0..test_geometry_vertices.len() {
            assert!(test_geometry_vertices[i].coords.relative_eq(
                &relaxed_geometry_vertices[i].coords,
                0.001,
                0.001,
            ));
        }
    }

    #[test]
    fn test_laplacian_smoothing_with_anchors_stop_when_stable_find_border_vertices() {
        let (faces, vertices) = shape_for_smoothing_with_anchors();
        let geometry = Geometry::from_triangle_faces_with_vertices_and_computed_normals(
            faces.clone(),
            vertices.clone(),
            NormalStrategy::Sharp,
        );

        let oriented_edges: Vec<OrientedEdge> = geometry.oriented_edges_iter().collect();
        let edge_sharing_map = edge_analysis::edge_sharing(&oriented_edges);
        let fixed_vertex_indices =
            Vec::from_iter(mesh_analysis::border_vertex_indices(&edge_sharing_map).into_iter());

        let (faces_correct, vertices_correct) = shape_for_smoothing_with_anchors_50_iterations();
        let test_geometry_correct =
            Geometry::from_triangle_faces_with_vertices_and_computed_normals(
                faces_correct.clone(),
                vertices_correct.clone(),
                NormalStrategy::Sharp,
            );

        let v2v = mesh_topology_analysis::vertex_to_vertex_topology(&geometry);
        let (relaxed_geometry, _, _) =
            laplacian_smoothing(&geometry, &v2v, 255, &fixed_vertex_indices, true);

        let relaxed_geometry_faces = relaxed_geometry.faces();
        let test_geometry_faces = test_geometry_correct.faces();

        assert_eq!(relaxed_geometry_faces, test_geometry_faces);

        let relaxed_geometry_vertices = relaxed_geometry.vertices();
        let test_geometry_vertices = test_geometry_correct.vertices();

        for i in 0..test_geometry_vertices.len() {
            assert!(test_geometry_vertices[i].coords.relative_eq(
                &relaxed_geometry_vertices[i].coords,
                0.001,
                0.001,
            ));
        }
    }
}
