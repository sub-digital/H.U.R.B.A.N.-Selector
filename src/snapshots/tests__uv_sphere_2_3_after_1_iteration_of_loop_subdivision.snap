---
source: src/mesh_smoothing.rs
expression: "&subdivided_geometry"
---
{
  "faces": [
    {
      "Triangle": {
        "vertices": [
          1,
          8,
          10
        ],
        "normals": [
          0,
          0,
          0
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          3,
          9,
          8
        ],
        "normals": [
          1,
          1,
          1
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          4,
          10,
          9
        ],
        "normals": [
          2,
          2,
          2
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          8,
          9,
          10
        ],
        "normals": [
          3,
          3,
          3
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          0,
          11,
          12
        ],
        "normals": [
          4,
          4,
          4
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          3,
          8,
          11
        ],
        "normals": [
          5,
          5,
          5
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          1,
          12,
          8
        ],
        "normals": [
          6,
          6,
          6
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          8,
          12,
          11
        ],
        "normals": [
          7,
          7,
          7
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          2,
          13,
          15
        ],
        "normals": [
          8,
          8,
          8
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          4,
          14,
          13
        ],
        "normals": [
          9,
          9,
          9
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          5,
          15,
          14
        ],
        "normals": [
          10,
          10,
          10
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          13,
          14,
          15
        ],
        "normals": [
          11,
          11,
          11
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          1,
          10,
          16
        ],
        "normals": [
          12,
          12,
          12
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          4,
          13,
          10
        ],
        "normals": [
          13,
          13,
          13
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          2,
          16,
          13
        ],
        "normals": [
          14,
          14,
          14
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          10,
          13,
          16
        ],
        "normals": [
          15,
          15,
          15
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          0,
          17,
          11
        ],
        "normals": [
          16,
          16,
          16
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          5,
          18,
          17
        ],
        "normals": [
          17,
          17,
          17
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          3,
          11,
          18
        ],
        "normals": [
          18,
          18,
          18
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          11,
          17,
          18
        ],
        "normals": [
          19,
          19,
          19
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          0,
          19,
          17
        ],
        "normals": [
          20,
          20,
          20
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          2,
          15,
          19
        ],
        "normals": [
          21,
          21,
          21
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          5,
          17,
          15
        ],
        "normals": [
          22,
          22,
          22
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          15,
          17,
          19
        ],
        "normals": [
          23,
          23,
          23
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          0,
          12,
          21
        ],
        "normals": [
          24,
          24,
          24
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          1,
          20,
          12
        ],
        "normals": [
          25,
          25,
          25
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          6,
          21,
          20
        ],
        "normals": [
          26,
          26,
          26
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          12,
          20,
          21
        ],
        "normals": [
          27,
          27,
          27
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          3,
          22,
          9
        ],
        "normals": [
          28,
          28,
          28
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          7,
          23,
          22
        ],
        "normals": [
          29,
          29,
          29
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          4,
          9,
          23
        ],
        "normals": [
          30,
          30,
          30
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          9,
          22,
          23
        ],
        "normals": [
          31,
          31,
          31
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          1,
          16,
          20
        ],
        "normals": [
          32,
          32,
          32
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          2,
          24,
          16
        ],
        "normals": [
          33,
          33,
          33
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          6,
          20,
          24
        ],
        "normals": [
          34,
          34,
          34
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          16,
          24,
          20
        ],
        "normals": [
          35,
          35,
          35
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          4,
          23,
          14
        ],
        "normals": [
          36,
          36,
          36
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          7,
          25,
          23
        ],
        "normals": [
          37,
          37,
          37
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          5,
          14,
          25
        ],
        "normals": [
          38,
          38,
          38
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          14,
          23,
          25
        ],
        "normals": [
          39,
          39,
          39
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          0,
          21,
          19
        ],
        "normals": [
          40,
          40,
          40
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          6,
          24,
          21
        ],
        "normals": [
          41,
          41,
          41
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          2,
          19,
          24
        ],
        "normals": [
          42,
          42,
          42
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          19,
          21,
          24
        ],
        "normals": [
          43,
          43,
          43
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          3,
          18,
          22
        ],
        "normals": [
          44,
          44,
          44
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          5,
          25,
          18
        ],
        "normals": [
          45,
          45,
          45
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          7,
          22,
          25
        ],
        "normals": [
          46,
          46,
          46
        ]
      }
    },
    {
      "Triangle": {
        "vertices": [
          18,
          25,
          22
        ],
        "normals": [
          47,
          47,
          47
        ]
      }
    }
  ],
  "vertices": [
    [
      0.50878996,
      -0.056250006,
      0.3875
    ],
    [
      -0.20568109,
      0.46874997,
      0.3875
    ],
    [
      -0.3031088,
      -0.41250002,
      0.3875
    ],
    [
      0.5087899,
      0.056250002,
      -0.38750005
    ],
    [
      -0.30310893,
      0.41249996,
      -0.38750005
    ],
    [
      -0.20568095,
      -0.46874997,
      -0.38750005
    ],
    [
      0.0,
      -1.4901161e-8,
      0.71875
    ],
    [
      0.0,
      0.0,
      -0.71875
    ],
    [
      0.2165063,
      0.375,
      -5.9604645e-8
    ],
    [
      0.10825314,
      0.375,
      -0.43750006
    ],
    [
      -0.27063298,
      0.46875,
      -5.9604645e-8
    ],
    [
      0.54126585,
      0.0,
      -5.9604645e-8
    ],
    [
      0.27063292,
      0.28125,
      0.43749994
    ],
    [
      -0.43301272,
      -2.9802322e-8,
      -5.9604645e-8
    ],
    [
      -0.3788861,
      -0.09375,
      -0.43750006
    ],
    [
      -0.27063286,
      -0.46875003,
      -5.9604645e-8
    ],
    [
      -0.37888613,
      0.09374997,
      0.43749994
    ],
    [
      0.2165064,
      -0.375,
      -5.9604645e-8
    ],
    [
      0.27063298,
      -0.28125,
      -0.43750006
    ],
    [
      0.10825324,
      -0.37500003,
      0.43749994
    ],
    [
      -0.108253196,
      0.1875,
      0.6875
    ],
    [
      0.21650636,
      -7.450581e-9,
      0.6875
    ],
    [
      0.21650635,
      0.0,
      -0.68750006
    ],
    [
      -0.10825319,
      0.1875,
      -0.6875
    ],
    [
      -0.10825316,
      -0.18750003,
      0.6875
    ],
    [
      -0.10825315,
      -0.1875,
      -0.6875
    ]
  ],
  "normals": [
    [
      0.03632813,
      0.1887665,
      -0.006089226
    ],
    [
      0.13945313,
      0.16982217,
      -0.034505695
    ],
    [
      0.011718733,
      0.16102658,
      -0.024356978
    ],
    [
      0.041015625,
      0.21312344,
      -0.010148734
    ],
    [
      0.13359377,
      0.09066207,
      0.024356946
    ],
    [
      0.14531249,
      0.12584431,
      0.006089244
    ],
    [
      0.07734374,
      0.20568106,
      0.0345057
    ],
    [
      0.1640625,
      0.1420823,
      0.010148725
    ],
    [
      -0.18164065,
      -0.0629222,
      -0.0060892324
    ],
    [
      -0.21679686,
      0.035858847,
      -0.03450571
    ],
    [
      -0.14531249,
      -0.07036459,
      -0.024356976
    ],
    [
      -0.20507812,
      -0.07104119,
      -0.010148746
    ],
    [
      -0.14531252,
      0.07036455,
      0.024356967
    ],
    [
      -0.18164062,
      0.06292215,
      0.006089235
    ],
    [
      -0.21679689,
      -0.035858907,
      0.034505717
    ],
    [
      -0.20507814,
      0.07104114,
      0.010148738
    ],
    [
      0.14531252,
      -0.12584428,
      -0.006089261
    ],
    [
      0.07734374,
      -0.20568103,
      -0.0345057
    ],
    [
      0.13359375,
      -0.09066201,
      -0.024356961
    ],
    [
      0.1640625,
      -0.14208224,
      -0.010148734
    ],
    [
      0.13945314,
      -0.16982216,
      0.034505695
    ],
    [
      0.011718752,
      -0.16102661,
      0.024356965
    ],
    [
      0.036328133,
      -0.18876646,
      0.0060892147
    ],
    [
      0.04101564,
      -0.21312344,
      0.010148726
    ],
    [
      0.09843751,
      0.056832947,
      0.08524939
    ],
    [
      0.042187512,
      0.13802283,
      0.11569557
    ],
    [
      0.005859375,
      0.010148736,
      0.040594947
    ],
    [
      0.046875007,
      0.08118991,
      0.10148735
    ],
    [
      0.0984375,
      0.105546854,
      -0.115695566
    ],
    [
      0.005859364,
      0.010148729,
      -0.04059494
    ],
    [
      -1.3038516e-8,
      0.113665804,
      -0.085249364
    ],
    [
      0.046874978,
      0.08118988,
      -0.101487346
    ],
    [
      -0.09843752,
      0.056832906,
      0.08524937
    ],
    [
      -0.14062503,
      -0.03247597,
      0.11569558
    ],
    [
      -0.011718751,
      -1.1641532e-9,
      0.040594943
    ],
    [
      -0.09375003,
      -1.4901161e-8,
      0.10148736
    ],
    [
      -0.14062496,
      0.032475937,
      -0.11569558
    ],
    [
      -0.01171875,
      -1.1641532e-9,
      -0.04059494
    ],
    [
      -0.09843747,
      -0.05683293,
      -0.085249364
    ],
    [
      -0.09374998,
      -7.450581e-9,
      -0.10148735
    ],
    [
      0.09843751,
      -0.105546854,
      0.115695596
    ],
    [
      0.005859376,
      -0.010148735,
      0.040594947
    ],
    [
      9.313226e-9,
      -0.11366585,
      0.08524938
    ],
    [
      0.04687502,
      -0.08118989,
      0.10148736
    ],
    [
      0.09843751,
      -0.056832895,
      -0.085249364
    ],
    [
      0.042187482,
      -0.13802277,
      -0.11569558
    ],
    [
      0.005859364,
      -0.010148728,
      -0.04059494
    ],
    [
      0.046874985,
      -0.08118988,
      -0.10148735
    ]
  ]
}
